webapp/javascript/components/TimelineChart/Annotations.plugin.tsx

```diff
@@ -0,0 +1,285 @@
+  content: string;
+};
+
+export interface ContextMenuProps {
+  click: {
+    /** The X position in the window where the click originated */
+    pageX: number;
+    /** The Y position in the window where the click originated */
+    pageY: number;
+  };
+  timestamp: number;
+  containerEl: HTMLElement;
+  value?: {
+    timestamp: number;
+    content: string;
+  } | null;
+}
+
+const WRAPPER_ID = randomId('contextMenu');
+
+const getIconByAnnotationType = (type: string) => {
+  switch (type) {
+    case 'message':
+    default:
+      return AnnotationMsgIcon;
+  }
+};
+
+const shouldStartAnnotationsFunctionality = (annotations?: AnnotationType[]) =>
+  Array.isArray(annotations);
+
+const inject = ($: JQueryStatic) => {
+  const alreadyInitialized = $(`#${WRAPPER_ID}`).length > 0;
+
+  if (alreadyInitialized) {
+    return $(`#${WRAPPER_ID}`);
+  }
+
+  const body = $('body');
+  return $(`<div id="${WRAPPER_ID}" />`).appendTo(body);
+};
+
+const getCursorPositionInPx = (
+  plot: jquery.flot.plot,
+  positionInTimestamp: { x: number; y: number }
+) => {
+  const axes = plot.getAxes();
+  const extractedX = extractRange(plot, axes, 'x');
+  const extractedY = extractRange(plot, axes, 'y');
+  const plotOffset = plot.getPlotOffset() as {
+    top: number;
+    left: number;
+  };
+
+  return {
+    x: Math.floor(extractedX.axis.p2c(positionInTimestamp.x)) + plotOffset.left,
+    y: Math.floor(extractedY.axis.p2c(positionInTimestamp.y)) + plotOffset.top,
+  };
+};
+
+const findAnnotationByCursorPosition = (
+  x: number,
+  y: number,
+  list: AnnotationPosition[] = []
+) => {
+  return list?.find((an) => {
+    return x >= an.fromX && x <= an.toX && y >= an.fromY && y <= an.toY;
+  });
+};
+
+function drawRoundRect(
+  ctx: CtxType,
+  x: number,
+  y: number,
+  w: number,
+  h: number,
+  radius: number,
+  color: string | Color
+) {
+  const r = x + w;
+  const b = y + h;
+  ctx.beginPath();
+  ctx.strokeStyle = color;
+  ctx.fillStyle = color;
+  ctx.lineWidth = 1;
+  ctx.moveTo(x + radius, y);
+  ctx.lineTo(r - radius, y);
+  ctx.quadraticCurveTo(r, y, r, y + radius);
+  ctx.lineTo(r, y + h - radius);
+  ctx.quadraticCurveTo(r, b, r - radius, b);
+  ctx.lineTo(x + radius, b);
+  ctx.quadraticCurveTo(x, b, x, b - radius);
+  ctx.lineTo(x, y + radius);
+  ctx.quadraticCurveTo(x, y, x + radius, y);
+  ctx.fill();
+  ctx.stroke();
+}
+
+(function ($) {
+  function init(plot: jquery.flot.plot & jquery.flot.plotOptions) {
+    const annotationsPositions: AnnotationPosition[] = [];
+    const placeholder = plot.getPlaceholder();
+
+    function onHover(_: unknown, pos: { x: number; y: number }) {
+      if (annotationsPositions?.length) {
+        const { x, y } = getCursorPositionInPx(plot, pos);
+
+        const annotation = findAnnotationByCursorPosition(
+          x,
+          y,
+          annotationsPositions
+        );
+
+        if (annotation) {
+          placeholder.trigger('hoveringOnAnnotation', [{ hovering: true }]);
```

Comment:
Dunno about this... message passing like this to a completely different plugin feels very hacky...

Reply:
removed